File: Path Sum II/Path Sum II.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @param sum, an integer
    # @return a list of lists of integers
    def pathSum(self, root, sum):
        stack=[]
        res=[]
        self.search(root, sum, stack, res)
        return res
    def search(self, root, sum, stack, res):
        if not root:
            return
        stack.append(root.val)
        if not root.left and not root.right:
            if root.val==sum:
                ans=[]
                for item in stack:
                    ans.append(item)
                res.append(ans)
        else:
            self.search(root.left, sum-root.val, stack, res)
            self.search(root.right, sum-root.val, stack, res)
        stack.pop()
                

File: Path Sum II/Path Sum II.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
     public ArrayList<ArrayList<Integer>> pathSum(TreeNode root, int sum) {
        ArrayList<Integer> stack = new ArrayList<Integer>();
        ArrayList<ArrayList<Integer>> res = new ArrayList<ArrayList<Integer>>();
        search(root, sum, stack, res);
        return res;
    }
    public void search(TreeNode root, int sum,ArrayList<Integer> stack, ArrayList<ArrayList<Integer>> res) {
        if (root == null) return;
        stack.add(root.val);
        if ((root.left == null) && (root.right == null)) {
            if (root.val == sum) {
                ArrayList<Integer> ans = new ArrayList<Integer>();
                for (Integer item: stack) ans.add(item);
                res.add(ans);
            }
        } else {
            search(root.left, sum - root.val, stack, res);
            search(root.right, sum - root.val, stack, res);
        }
        stack.remove(stack.size() - 1);
    }

}
